### PR TITLE
[BatchMode] Preserve intra-batch order, add a seed mode.

### PR DESCRIPTION
**Pull request by graydon**
This contains 3 changes, all sufficiently related that I didn't bother splitting them.

  1. Use SetVectors in places where we were using Sets, so input file order is preserved.
  2. Simplify batching so there's no straggler job.
  3. Add a `-driver-batch-seed` flag you can use to randomize the batch partition.

Pleasantly, I also realized `-driver-skip-execution` means that I can actually test the driver batching code in isolation, so I added a test for that too :)

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
cc @davidungar

File: lib/Driver/Compilation.cpp [hunk 1]
```diff
@@ -120,7 +126,8 @@ Compilation::Compilation(DiagnosticEngine &Diags,
 static bool writeFilelistIfNecessary(const Job *job, DiagnosticEngine &diags);
 
 using CommandSet = llvm::SmallPtrSet<const Job *, 16>;
-
+using CommandSetVec = llvm::SetVector<const Job*>;
+using BatchPartition = std::vector<std::unique_ptr<CommandSetVec>>;
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Why the indirection here?

File: lib/Driver/Compilation.cpp [hunk 2]
```diff
@@ -120,7 +126,8 @@ Compilation::Compilation(DiagnosticEngine &Diags,
 static bool writeFilelistIfNecessary(const Job *job, DiagnosticEngine &diags);
 
 using CommandSet = llvm::SmallPtrSet<const Job *, 16>;
-
+using CommandSetVec = llvm::SetVector<const Job*>;
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Nitpick: I don't think the "Vec" abbreviation is buying anything here. "Vector" is fine.

File: test/Driver/batch_mode.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// RUN: %empty-directory(%t)
+// RUN: touch %t/file-01.swift %t/file-02.swift %t/file-03.swift %t/file-04.swift %t/file-05.swift
+// RUN: touch %t/file-06.swift %t/file-07.swift %t/file-08.swift %t/file-09.swift %t/file-10.swift
+// RUN: touch %t/file-11.swift %t/file-12.swift %t/file-13.swift %t/file-14.swift %t/file-15.swift
+// RUN: touch %t/file-16.swift %t/file-17.swift %t/file-18.swift %t/file-19.swift %t/file-20.swift
+// RUN: touch %t/file-21.swift %t/file-22.swift %t/file-23.swift %t/file-24.swift %t/file-25.swift
+// RUN: touch %t/file-26.swift %t/file-27.swift %t/file-28.swift %t/file-29.swift %t/file-30.swift
+// RUN: %swiftc_driver -enable-batch-mode -driver-show-job-lifecycle -driver-skip-execution -j 10 %t/file-*.swift | %FileCheck %s -check-prefix=SEED0
```
Comment (jrose-apple) [inline test/Driver/batch_mode.swift]: Nitpick: I wouldn't want to rely on `*` producing files in either alphabetical or creation order. You should probably just list them out.

**Comment by jrose-apple** [comment]
I'm not sure I care as much about round-robin. It also makes it harder for people to manually tweak the batching if they need to in the early days. But the SetVector sounds important, and the randomization seed is good.

**Comment by graydon** [comment]
I don't get what you mean by harder to manually tweak; but the main point of the round-robin (or even just the first iteration of this patch, that funnelled everything through the prng and just seeded it to zero by default so it always did the same thing) is that most of the time -- unless NCORES perfectly divides the size of the input set -- users of the previous iteration would wind up running NCORES full-sized jobs _plus one straggler job_ with a 1..NCORES-1-job remainder batch (which by definition there aren't enough cores to run in parallel, so then we're throwing extra time-slicing in while the straggler runs).

(The "less clumping" term is a more hypothetical notion: that it's possible a set of source files fed to the compiler have some natural asymmetry in workload that corresponds to filenames -- eg. a main directory full of expensive model files with many large methods and a secondary directory full of less-expensive utility files with small helpers -- and so spreading filename lexicographic order across jobs might also help out that workload variability. Very much more a guess though; equally possible that doing so makes each batch job do "more" overall work because it defeats locality in, say, the types each job lazily validates. I don't actually think this will be much of a factor either way, but several people have expressed concern when we discussed static scheduling that there would be some clumping that makes a simple "1/N partition" unbalanced.)

File: lib/Driver/Compilation.cpp (hunk 2 above)
Comment (graydon) [reply lib/Driver/Compilation.cpp]: ok

File: lib/Driver/Compilation.cpp (hunk 1 above)
Comment (graydon) [reply lib/Driver/Compilation.cpp]: Hm I think this is residue from an iteration of the patch that paid more attention to partition-element identity. Will remove if so, good catch.

File: test/Driver/batch_mode.swift (hunk 3 above)
Comment (graydon) [reply test/Driver/batch_mode.swift]: Eek! google "Python glob" => "results are returned in arbitrary order", indeed.

**Comment by graydon** [comment]
@jrose-apple pushed update with review comments addressed.

**Comment by graydon** [comment]
@swift-ci please test

**Comment by jrose-apple** [comment]
If we want to get fancy about filenames, files with similar prefixes are probably part of the same subsystem and therefore would benefit more from being in the same batch. But the input order isn't actually alphabetical anyway.

It's good to solve the straggler job problem, but this isn't the only way to do it. I'm pretty sure simply rounding up to the next multiple of NCORES would accomplish the same thing.

File: test/Driver/batch_mode.swift (hunk 3 above)
Comment (jrose-apple) [reply test/Driver/batch_mode.swift]: I don't think we guarantee Python execution either. lit.py has a mode where it just immediately shoves everything into a shell, and I *think* that's the default on both macOS and Linux.

**Comment by graydon** [comment]
True! The actual history of how I got here was: "well we need to pre-allocate the partition to size N to solve stragglers and support randomization, and as a bonus I can delete reasoning about target batch size and just add jobs at random", then "ok I have a version that does randomization but I _bet_ it'll be undesirable for the default to be a weird-looking order, so I guess round-robin gets something predictable looking and addresses that bunching thing people were worried about"

But it's only another couple lines to reinstate a target of ceil(NJOBS/NCORES), so I'll do that :)

**Comment by graydon** [comment]
@jrose-apple updated w/ even spread of remainder across stable partition-order (and a slight improvement to logging).

**Comment by graydon** [comment]
@swift-ci please test

File: lib/Driver/Compilation.cpp [hunk 4]
```diff
@@ -714,45 +721,81 @@ namespace driver {
...
-                     size_t TargetBatchSize) {
-      if (!canExpandBatch(Cmd, CurrentBatch, TargetBatchSize)) {
-        formBatchJobFromCurrentBatch(Batches, CurrentBatch);
+    /// Shuffle \p Batchable if -driver-batch-seed is nonzero.
+    void maybeShuffleBatchable(CommandSetVector &Batchable) {
+      if (Comp.BatchSeed != 0) {
+        std::minstd_rand gen(Comp.BatchSeed);
+        auto tmp = Batchable.takeVector();
+        std::shuffle(tmp.begin(), tmp.end(), gen);
+        Batchable = CommandSetVector(tmp.begin(), tmp.end());
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Nitpick: maybe this function should just return a `std::vector` (and accept `Batchable` by value, and `partitionIntoBatches` can do the same) so as not to pay the cost of re-uniqueing the elements. It's only used as an array from here on out.

File: lib/Driver/Compilation.cpp [hunk 5]
```diff
@@ -714,45 +721,81 @@ namespace driver {
...
-                        llvm::SetVector<const Job *> &CurrentBatch,
-                        size_t TargetBatchSize) {
-      auto const &TC = Comp.getToolChain();
-      return (CurrentBatch.empty() ||
-              (TC.jobsAreBatchCombinable(Comp, Cmd, CurrentBatch[0]) &&
-               CurrentBatch.size() < TargetBatchSize));
+    /// Given 0 <= \p i < \p P.size(), advance i to the "next" partition,
+    /// which means filling each partition in order (distributing
+    /// any division-remainder evenly between partitions).
+    void nextPartition(size_t &i,
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: This function probably needs a new name, since it doesn't *always* advance `i` and `Partition`.

**Review by jrose-apple — APPROVED** [review]

File: lib/Driver/Compilation.cpp (hunk 4 above)
Comment (graydon) [reply lib/Driver/Compilation.cpp]: Oh, yeah, actually there are a bunch of needless setvector cases after the initial split that should just be vectors. Good catch, fixing.

File: lib/Driver/Compilation.cpp (hunk 5 above)
Comment (graydon) [reply lib/Driver/Compilation.cpp]: right!

**Comment by graydon** [comment]
@swift-ci please test and merge

**Comment by graydon** [comment]
@swift-ci please test and merge

**Pull request merged**